shared/sbl.js

```diff
@@ -0,0 +1,111 @@
+  domainFilters: Array<TypeFilter>,
+  selectorFilters: {[string]: TypeFilter},
+  data: Object
+}
+
+var domains: {[string]: TypeDomain} = {}
+var globalFilters: Array<TypeFilter> = []
+var pendingCalls = {} // allows sbl to be called regardless of module load order
+
+const DOMAIN_SELECTOR_REGEX = /^([^/]+)(.+)/
+
+function sbl (path: string, ...data?: Array<*>) {
+  var [, name, selector] = DOMAIN_SELECTOR_REGEX.exec(path)
+  var domain = domains[name]
+  // Filters can perform additional functions, and by returning `false` they
+  // can prevent the execution of a selector.
+  for (let gf of globalFilters) {
+    if (gf(name, selector, data) === false) return
+  }
+  for (let df of domain.domainFilters) {
+    if (df(name, selector, data) === false) return
+  }
+  var sf = domain.selectorFilters[selector]
+  if (sf && sf(name, selector, data) === false) return
+  return domain.selectors[selector].call(domain, ...data)
+}
+
+sbl.registerDomain = function (domain: string, selectors: {[string]: Function}) {
+  if (domains[domain]) throw new Error(`${domain} exists`)
+  domains[domain] = {selectors, data: {}, domainFilters: [], selectorFilters: {}}
+  // run any pendingCalls for this domain
+  if (pendingCalls[domain]) {
+    for (var [path, data] of pendingCalls[domain]) {
+      sbl(path, ...data)
+    }
+    delete pendingCalls[domain]
+  }
+}
+
+// During startup of app, use this instead of sbl function to call selectors
+// on domains that haven't been registered yet. The selector gets run immediately
+// once the corresponding domain is registered via registerDomain.
+sbl.ready = function (path: string, ...data?: Array<*>) {
+  // a call might happen to a selector for domains that haven't been registered
+  // yet, due to module load order. So we save them and run them later.
+  var [, domain] = DOMAIN_SELECTOR_REGEX.exec(path)
+  if (!pendingCalls[domain]) pendingCalls[domain] = []
+  pendingCalls[domain].push([path, data])
+}
+
+sbl.addDomainFilter = function (domain: string, filter: TypeFilter) {
+  domains[domain].domainFilters.push(filter)
+}
+
+sbl.setSelectorFilter = function (domain: string, selector: string, filter: TypeFilter) {
+  if (domains[domain].selectorFilters[selector]) {
+    throw new Error(`${domain}/${selector} filter exists`)
+  }
+  domains[domain].selectorFilters[selector] = filter
+}
+
+sbl.addGlobalFilter = function (filter: TypeFilter) {
+  globalFilters.push(filter)
+}
+
+export default sbl
+
+// =======================
+// Common selectors APIs
+// TODO: move to a separate file
+// =======================
+
+function addListener (event: string, handler: Function, type: 'listeners' | 'listenOnce') {
+  if (!this.data.events) {
+    this.data.events = {}
+  }
+  if (!this.data.events[event]) {
+    this.data.events[event] = {listeners: [], listenOnce: []}
+  }
+  this.data.events[event][type].push(handler)
+}
+
+const COMMON_MIXINS = {
+  V1: {
+    EVENTS: {
+      '/v1/events/on': function (event, data) {
```

Comment:
A `TODO` as a reminder would help so no one uses it on new functionality expecting listners to clean themselves up